h2/src/main/org/h2/mvstore/tx/TransactionMap.java

```diff
@@ -216,14 +242,39 @@ public V putCommitted(K key, V value) {
     }
 
     private V set(K key, V value) {
-        transaction.checkNotClosed();
-        V old = get(key);
-        boolean ok = trySet(key, value, false);
-        if (ok) {
-            return old;
-        }
-        throw DataUtils.newIllegalStateException(
-                DataUtils.ERROR_TRANSACTION_LOCKED, "Entry is locked");
+        TxDecisionMaker decisionMaker = new TxDecisionMaker.PutDecisionMaker(map.getId(), key, value, transaction);
+        return set(key, decisionMaker);
+    }
+
+    @SuppressWarnings("unchecked")
+    private V set(K key, TxDecisionMaker decisionMaker) {
+        TransactionStore store = transaction.store;
+        Transaction blockingTransaction;
+        long sequenceNumWhenStarted;
+        VersionedValue result;
+        do {
+            sequenceNumWhenStarted = store.openTransactions.get().getVersion();
+            assert transaction.getBlockerId() == 0;
+
+            result = map.put(key, VersionedValue.DUMMY, decisionMaker);
```

Comment:
I'm not sure that I correctly understood the logic with this substituted value. May be some comments can be placed here?